[FLINK-24352] [flink-table-planner] Add null check for temporal table check on SqlSnapshot

YuvalItzchakov opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Purpose of this change is to resolve https://issues.apache.org/jira/browse/FLINK-24352 and it's duplicate https://issues.apache.org/jira/browse/FLINK-24956.

The below test which uses a lookup table join with a CTE fails due to a `NullPointerException`

```scala
package foo.bar

import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.table.api.DataTypes
import org.apache.flink.table.api.Schema
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
object Test {
  final case class Person(name: String, age: Int)

  def main(args: Array[String]): Unit = {
    val ee = StreamExecutionEnvironment.getExecutionEnvironment
    val te = StreamTableEnvironment.create(ee)    
    val personSchema = Schema.newBuilder().column("name", DataTypes.STRING()).column("age", DataTypes.INT()).build()
    val x = ee.fromCollection(List(Person("a", 1)))
    te.createTemporaryView(
      "my_table",
      x,
      personSchema
    )
    val y = ee.fromCollection(List(Person("b", 2)))
    te.createTemporaryView(
      "my_table_2",
      y,
      personSchema
    )
    val res =
      te.executeSql("""
                      |WITH A AS (
                      |  select name, age + 1 from my_table
                      |),
                      |B AS (
                      |  select name, age + 2 from my_table_2
                      |)
                      |
                      |SELECT A.name, B.age
                      |FROM A
                      |JOIN B
                      |FOR SYSTEM_TIME AS OF PROCTIME() on (A.name = B.name)
                      |""".stripMargin)
    res.print()
  }
}
```

@tsreaper waiting on your comments. I'm not entirely sure a null check is sufficient to determine the underlying table is actually temporal.

## Brief change log

- Add null check for `SqlValidatorImpl.validateSnapshot` before checking the underling table, which may be null in the case the expression is a CTE.

## Verifying this change

This change added tests and can be verified as follows:

- Added a test to TableEnvironmentITCase which validates calling a lookup join on a CTE no longer throws a NPE.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

tsreaper reviewed (changes requested):
Hi @YuvalItzchakov , thanks for your interest in contributing to Flink. However, this PR has some major issues.

1. Please add the corresponding issue id and name of the module at the beginning of your commit message and the title of your PR. You can also check other PRs.
2. Current solution is too broad. `ns.getTable` will return null by default if a subclass of `AbstractNamespace` does not override this method, so the current solution will allow almost anything to pass. Can it be more specific? For example, check if `ns` is an `IdentifierNamespace` which resolves to a `WithItemNamespace`.
3. If you want to validate SQL plans in a test, add them to test classes like `LookupJoinTest` instead of IT case classes.

YuvalItzchakov commented:
@tsreaper I have fixed the PR name and made the check more specific. Additionally I replaced the test ITCase test with a test in LookupJoinTest as you mentioned.

tsreaper commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/join/LookupJoinTest.scala:
```diff
@@ -525,7 +525,21 @@ class LookupJoinTest(legacyTableSource: Boolean) extends TableTestBase with Seri
+        |ON MyLookupTable.a = D.id AND D.age = 10
+      """.stripMargin
+
+    verifyTranslationSuccess(sql)
```
I would recommend `util.verifyExecPlan` so that we can compare the exact plan in the future.

Note that to use this checking method, you'll have to first remove the corresponding LookupJoinTest.xml file of this test then run the whole test class. You'll find that a new LookupJoinTest.xml is generated and then git will show you what is added to the new LookupJoinTest.xml. Make sure that the added plan meets your expectation.

YuvalItzchakov replied:
@tsreaper I can add an additional test that verifies to exec plan, as it was similar before and after my change (I originally tried using it in my test, but the `NullPointerException` did not reproduce using it.

tsreaper replied:
@YuvalItzchakov I can reproduce the `NullPointerException` with `util.verifyExecPlan`. Maybe you forget to revert the changes before testing.

```suggestion
    util.verifyExecPlan(sql)
```

tsreaper commented:
cc. @godfreyhe do you have any other thoughts about this solution?

YuvalItzchakov commented:
> cc. @godfreyhe do you have any other thoughts about this solution?

@godfreyhe Would appreciate any pointers.

godfreyhe commented:
@YuvalItzchakov Thanks for the contribution, I'm be fine with the current solution even it not very well, since we do not check whether the query in `with` is temporal (just like the following branch you added). That do not affect the result because we will check it again in planner rules.

MartijnVisser commented:
Should we create a follow-up ticket for the issue you've mentioned @godfreyhe ?

tsreaper commented:
@godfreyhe 

> since we do not check whether the query in `with` is temporal

I don't quite understand this. I think `with` is always temporal?

YuvalItzchakov commented:
@godfreyhe @tsreaper I have fixed the PR comments. Are we good with merging this?

tsreaper reviewed (approved):
Looks good to me. @godfreyhe for merging if you're also OK with this.

YuvalItzchakov commented:
@godfreyhe Would appreciate your reply so we can merge this into 1.14.1 as well

YuvalItzchakov commented:
Anyone?

MartijnVisser commented:
Let me see if I can find someone who can help

wenlong88 commented:
hi, sorry for jumping in, but I think current fix didn't cover the other case that the getTable can be null. If there are sometime the getTable is null but can not be ignored in validation, we should throw an helpful exception instead of NPE.

tsreaper commented:
> hi, sorry for jumping in, but I think current fix didn't cover the other case that the getTable can be null. If there are sometime the getTable is null but can not be ignored in validation, we should throw an helpful exception instead of NPE.

@wenlong88 I don't get this. This PR is meant to fix the case when `getTable()` returns null. Could you elaborate more on this and preferably give us an counter-example?

wenlong88 commented on flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:
```diff
@@ -4711,6 +4711,11 @@ private void validateSnapshot(SqlNode node, SqlValidatorScope scope, SqlValidato
+            if (ns instanceof IdentifierNamespace && ns.resolve() instanceof WithItemNamespace) {
+                // If the snapshot is used over a CTE, then we don't have a concrete underlying
+                // table to operate on. This will be rechecked later in the planner rules.
+                return;
```
@tsreaper the fix here just ignore WithItemNamespace. If the ns is not WithItemNamespace, and ns.getTable is null(it may be possible according to your former comment), the validator will still throw NPE exception. 
I think it may be enough to  just ignore when getTable is null(this is  how the validator treat ns.getTable in other method), or it is better to throw a more helpful exception instead of NPE.

tsreaper replied:
I understand. If with this fix we still meet a case where `getTable()` is null then this is unexpected and we should take another look. So we should add a `Preconditions.notNull()` here with a message asking users to create a ticket.

godfreyhe replied:
@wenlong88 and @tsreaper Currently, The table will not be null, because the namespace for snapshot will always be IdentifierNamespace (snapshot must be with a table name, view name or the CTE name).`ns.resolve()` will check `whether resolvedNamespace is null`. But I'm also fine with we add `notNull` check to protect the future change. cc @YuvalItzchakov

YuvalItzchakov replied:
@godfreyhe What should happen in case it's null? Which exception should be propagated upwards?

godfreyhe replied:
@YuvalItzchakov Y I think you can throw `CalciteContextException` directly, just call `newValidationError()`

godfreyhe replied:
@YuvalItzchakov do you want to add null check ?

YuvalItzchakov commented:
@tsreaper @godfreyhe @MartijnVisser Guys this has been waiting for a month now although the PR is ready. 

This fix is affecting the ability of my consumers to use Flink and I'd really like to get this merged. 

As there is no more critical issues re table nullability, is there anything else stalling us?

godfreyhe commented:
@YuvalItzchakov I add `assert ns.getTable() != null`  before calling `ns.getTable().isTemporal()` and merge the pr. Thanks for the contribution
